clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -633,37 +633,110 @@ static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef InputTable,
   return *Output;
 }
 
-// Run clang-offload-wrapper
-static Expected<StringRef> runWrapper(StringRef &InputFile,
-                                      const ArgList &Args) {
-  // Create a new file to write the wrapped file to.
-  auto TempFileOrErr =
-      createOutputFile(sys::path::filename(ExecutableName), "bc");
-  if (!TempFileOrErr)
-    return TempFileOrErr.takeError();
-  Expected<std::string> ClangOffloadWrapperPath = findProgram(
-      "clang-offload-wrapper", {getMainExecutable("clang-offload-wrapper")});
-  if (!ClangOffloadWrapperPath)
-    return ClangOffloadWrapperPath.takeError();
+Expected<std::string> readFile(StringRef File) {
+  auto MBOrErr = MemoryBuffer::getFile(File);
+  if (!MBOrErr)
+    return createFileError(File, MBOrErr.getError());
 
-  BumpPtrAllocator Alloc;
-  StringSaver Saver(Alloc);
+  auto &MB = *MBOrErr;
+  return std::string(MB->getBufferStart(), MB->getBufferEnd());
+}
+
+Expected<std::unique_ptr<util::PropertySetRegistry>>
+readPropertyRegistryFromFile(StringRef File) {
+  auto MBOrErr = MemoryBuffer::getFile(File);
+  if (!MBOrErr)
+    return createFileError(File, MBOrErr.getError());
+
+  auto &MB = *MBOrErr;
+  return util::PropertySetRegistry::read(&*MB);
+}
+
+Expected<SmallVector<SYCLImage>> readImages(StringRef File) {
+  auto MBOrErr = MemoryBuffer::getFile(File);
+  if (!MBOrErr)
+    return createFileError(File, MBOrErr.getError());
+
+  line_iterator LI(**MBOrErr);
+  // check the header.
+  if (LI.is_at_eof() || *LI != "[Code|Properties|Symbols]")
+    return createStringError(inconvertibleErrorCode(),
+                             "invalid SYCL Table file.");
+
+  ++LI;
+  SmallVector<SYCLImage> Images;
+  for (; !LI.is_at_eof(); ++LI) {
+    StringRef Line = *LI;
+    SmallVector<StringRef, 3> Elems;
+    Line.split(Elems, '|'); // expected elems are Code, Properties and Symbols.
+
+    if (Elems.size() != 3)
+      return createStringError(inconvertibleErrorCode(),
+                               "invalid SYCL Table file.");
+
+    auto ImageOrErr = readFile(Elems[0]);
+    if (!ImageOrErr)
+      return ImageOrErr.takeError();
+
+    auto PropertiesOrErr = readPropertyRegistryFromFile(Elems[1]);
+    if (!PropertiesOrErr)
+      return PropertiesOrErr.takeError();
+
+    auto SymbolsOrErr = readFile(Elems[2]);
+    if (!SymbolsOrErr)
+      return SymbolsOrErr.takeError();
+
+    SYCLImage Image;
+    Image.Image = std::move(*ImageOrErr);
+    Image.PropertyRegistry = std::move(**PropertiesOrErr);
+    Image.Entries = std::move(*SymbolsOrErr);
+    Images.push_back(std::move(Image));
+    // TODO: where are compile-opts, link-opts and other stuff?
```

Comment:
We can remove compile-opts and link-opts. I do not see them being used in SYCL compilation flow. @mdtoguchi for comment here.

Reply:
`-link-opts` and `-compile-opts` as passed to clang-offload-wrapper are populated with `-Xsycl-target-backend` values or `-Xsycl-target-linker` values.

Reply:
So `-compile-opts` is being used. @mdtoguchi Do you know whether `-Xsycl-target-linker` is actually being used by somebody? I can't come up with a corresponding user case.

Reply:
I don't know of any actual users of the option.

Reply:
I can see that there is one command line option for linker options
https://github.com/intel/llvm/blob/ab83849773d0cfdee3d2622b537ed26e8275a07d/clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td#L83

But it is unclear whether is it ok to use this for getting content from `-Xsycl-target-linker`.
I will add one additional command line option for `-compile-opts`. Also I added a question regarding purpose of `-Xsycl-target-linker` in internal communication.